src/mongo/crypto/crypto_openssl.cpp

```diff
@@ -39,25 +39,46 @@
 #include <openssl/hmac.h>
 #include <openssl/sha.h>
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+namespace {
+// Copies of OpenSSL after 1.1.0 define new EVP digest routines. We must
+// polyfill used definitions to interact with older OpenSSL versions.
+#include <cstring>
```

Comment:
One last change: This include needs to be outside of the anonymous namespace. This causes compiler failures on some platforms that look like:

> [2017/01/18 16:15:12.055] In file included from src/mongo/crypto/crypto_openssl.cpp:46:0:
[2017/01/18 16:15:12.055] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:75:11: error: '::memchr' has not been declared
[2017/01/18 16:15:12.055]    using ::memchr;
[2017/01/18 16:15:12.055]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:77:11: error: '::memcpy' has not been declared
[2017/01/18 16:15:12.056]    using ::memcpy;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:78:11: error: '::memmove' has not been declared
[2017/01/18 16:15:12.056]    using ::memmove;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:79:11: error: '::memset' has not been declared
[2017/01/18 16:15:12.056]    using ::memset;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:80:11: error: '::strcat' has not been declared
[2017/01/18 16:15:12.056]    using ::strcat;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:81:11: error: '::strcmp' has not been declared
[2017/01/18 16:15:12.056]    using ::strcmp;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:83:11: error: '::strcpy' has not been declared
[2017/01/18 16:15:12.056]    using ::strcpy;
[2017/01/18 16:15:12.056]            ^
[2017/01/18 16:15:12.056] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:84:11: error: '::strcspn' has not been declared
[2017/01/18 16:15:12.057]    using ::strcspn;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:86:11: error: '::strlen' has not been declared
[2017/01/18 16:15:12.057]    using ::strlen;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:87:11: error: '::strncat' has not been declared
[2017/01/18 16:15:12.057]    using ::strncat;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:88:11: error: '::strncmp' has not been declared
[2017/01/18 16:15:12.057]    using ::strncmp;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:89:11: error: '::strncpy' has not been declared
[2017/01/18 16:15:12.057]    using ::strncpy;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:90:11: error: '::strspn' has not been declared
[2017/01/18 16:15:12.057]    using ::strspn;
[2017/01/18 16:15:12.057]            ^
[2017/01/18 16:15:12.057] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:93:11: error: '::strchr' has not been declared
[2017/01/18 16:15:12.058]    using ::strchr;
[2017/01/18 16:15:12.058]            ^
[2017/01/18 16:15:12.058] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:94:11: error: '::strpbrk' has not been declared
[2017/01/18 16:15:12.058]    using ::strpbrk;
[2017/01/18 16:15:12.058]            ^
[2017/01/18 16:15:12.058] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:95:11: error: '::strrchr' has not been declared
[2017/01/18 16:15:12.058]    using ::strrchr;
[2017/01/18 16:15:12.058]            ^
[2017/01/18 16:15:12.058] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:96:11: error: '::strstr' has not been declared
[2017/01/18 16:15:12.058]    using ::strstr;
[2017/01/18 16:15:12.058]            ^
[2017/01/18 16:15:12.058] src/mongo/crypto/crypto_openssl.cpp: In function 'bool mongo::crypto::sha1(const unsigned char*, size_t, unsigned char*)':
[2017/01/18 16:15:12.059] src/mongo/crypto/crypto_openssl.cpp:70:5: error: reference to 'std' is ambiguous
[2017/01/18 16:15:12.059]      std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)> digestCtx(EVP_MD_CTX_new(),
[2017/01/18 16:15:12.059]      ^
[2017/01/18 16:15:12.059] <built-in>: note: candidates are: namespace std { }
[2017/01/18 16:15:12.059] In file included from src/mongo/crypto/crypto_openssl.cpp:46:0:
[2017/01/18 16:15:12.059] /opt/mongodbtoolchain/v2/include/c++/5.4.0/cstring:72:1: note:                 namespace <unnamed>::std { }
[2017/01/18 16:15:12.059]  {
[2017/01/18 16:15:12.059]  ^
[2017/01/18 16:15:12.059] src/mongo/crypto/crypto_openssl.cpp:70:31: error: expected primary-expression before ',' token
[2017/01/18 16:15:12.059]      std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)> digestCtx(EVP_MD_CTX_new(),
[2017/01/18 16:15:12.059]                                ^
[2017/01/18 16:15:12.059] src/mongo/crypto/crypto_openssl.cpp:70:33: error: expected primary-expression before 'decltype'
[2017/01/18 16:15:12.059]      std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)> digestCtx(EVP_MD_CTX_new(),
[2017/01/18 16:15:12.059]                                  ^
[2017/01/18 16:15:12.059] src/mongo/crypto/crypto_openssl.cpp:73:32: error: 'digestCtx' was not declared in this scope
[2017/01/18 16:15:12.059]      if (1 != EVP_DigestInit_ex(digestCtx.get(), EVP_sha1(), NULL)) {
[2017/01/18 16:15:12.059]                                 ^
[2017/01/18 16:15:12.060] src/mongo/crypto/crypto_openssl.cpp:77:31: error: 'digestCtx' was not declared in this scope
[2017/01/18 16:15:12.060]      if (1 != EVP_DigestUpdate(digestCtx.get(), input, inputLen)) {
[2017/01/18 16:15:12.060]                                ^
[2017/01/18 16:15:12.060] src/mongo/crypto/crypto_openssl.cpp:81:37: error: 'digestCtx' was not declared in this scope
[2017/01/18 16:15:12.060]      return (1 == EVP_DigestFinal_ex(digestCtx.get(), output, NULL));
[2017/01/18 16:15:12.060]                                      ^